#include "SelectDatabaseDialog.h"
#include "ui_SelectDatabaseDialog.h"

#include <QMessageBox>
#include <QSettings>

namespace SC::UI::Dialogs {

SelectDatabaseDialog::SelectDatabaseDialog(QWidget *parent)
    : QDialog(parent)
    , ui(new Ui::SelectDatabaseDialog)
    , m_model(new QStandardItemModel(this))
{
    ui->setupUi(this);    

    setupModel();
    loadConnections();

    ui->tableViewDatabases->setModel(m_model);

    // üîπ –ü–æ–∫–∞–∑—É—î–º–æ –ª–∏—à–µ —Ä—è–¥–∫–∏
    ui->tableViewDatabases->setSelectionBehavior(QAbstractItemView::SelectRows);
    ui->tableViewDatabases->setSelectionMode(QAbstractItemView::SingleSelection);

    // üîπ –ü—Ä–∏–±–∏—Ä–∞—î–º–æ –∑–∞–≥–æ–ª–æ–≤–∫–∏
    ui->tableViewDatabases->horizontalHeader()->hide();
    ui->tableViewDatabases->verticalHeader()->hide();

    // üîπ –•–æ–≤–∞—î–º–æ –≤—Å—ñ –∫–æ–ª–æ–Ω–∫–∏, –∫—Ä—ñ–º –ø–µ—Ä—à–æ—ó
    for (int col = 1; col < m_model->columnCount(); ++col) {
        ui->tableViewDatabases->setColumnHidden(col, true);
    }

    // üîπ –ü–µ—Ä—à–∞ –∫–æ–ª–æ–Ω–∫–∞ –Ω–∞ –≤—Å—é —à–∏—Ä–∏–Ω—É
    ui->tableViewDatabases->horizontalHeader()
        ->setSectionResizeMode(0, QHeaderView::Stretch);

    // üîπ –ë–µ–∑ —Å—ñ—Ç–∫–∏ (–≤–∏–≥–ª—è–¥–∞—î —è–∫ —Å–ø–∏—Å–æ–∫)
    ui->tableViewDatabases->setShowGrid(false);

    // üîπ –û–ø—Ü—ñ–æ–Ω–∞–ª—å–Ω–æ: –±–µ–∑ —Ä–∞–º–∫–∏
    ui->tableViewDatabases->setFrameShape(QFrame::NoFrame);

    connect(ui->btnAdd,    &QPushButton::clicked, this, &SelectDatabaseDialog::addDatabase);
    connect(ui->btnEdit,   &QPushButton::clicked, this, &SelectDatabaseDialog::editDatabase);
    connect(ui->btnRemove, &QPushButton::clicked, this, &SelectDatabaseDialog::removeDatabase);
    connect(ui->btnSelect, &QPushButton::clicked, this, &SelectDatabaseDialog::selectDatabase);
    connect(ui->btnCancel, &QPushButton::clicked, this, &QDialog::reject);
}

SelectDatabaseDialog::~SelectDatabaseDialog()
{
    delete ui;
}

/* =========================
 *  Public
 * ========================= */

DatabaseConnectionInfo SelectDatabaseDialog::selectedDatabase() const
{
    DatabaseConnectionInfo info;

    if (m_selectedRow < 0)
        return info;

    info.name     = m_model->item(m_selectedRow, 0)->text();
    info.host     = m_model->item(m_selectedRow, 1)->text();
    info.port     = m_model->item(m_selectedRow, 2)->text().toInt();
    info.database = m_model->item(m_selectedRow, 3)->text();
    info.user     = m_model->item(m_selectedRow, 4)->text();

    return info;
}

/* =========================
 *  Slots
 * ========================= */

void SelectDatabaseDialog::addDatabase()
{
    QMessageBox::information(this, tr("–î–æ–¥–∞—Ç–∏"),
                             tr("–î—ñ–∞–ª–æ–≥ –¥–æ–¥–∞–≤–∞–Ω–Ω—è –ë–î —â–µ –Ω–µ —Ä–µ–∞–ª—ñ–∑–æ–≤–∞–Ω–∏–π"));
}

void SelectDatabaseDialog::editDatabase()
{
    QMessageBox::information(this, tr("–†–µ–¥–∞–≥—É–≤–∞—Ç–∏"),
                             tr("–î—ñ–∞–ª–æ–≥ —Ä–µ–¥–∞–≥—É–≤–∞–Ω–Ω—è –ë–î —â–µ –Ω–µ —Ä–µ–∞–ª—ñ–∑–æ–≤–∞–Ω–∏–π"));
}

void SelectDatabaseDialog::removeDatabase()
{
    auto index = ui->tableViewDatabases->currentIndex();
    if (!index.isValid())
        return;

    m_model->removeRow(index.row());
    saveConnections();
}

void SelectDatabaseDialog::selectDatabase()
{
    auto index = ui->tableViewDatabases->currentIndex();
    if (!index.isValid()) {
        QMessageBox::warning(this, tr("–ü–æ–º–∏–ª–∫–∞"),
                             tr("–û–±–µ—Ä—ñ—Ç—å –±–∞–∑—É –¥–∞–Ω–∏—Ö"));
        return;
    }

    m_selectedRow = index.row();
    accept();
}

/* =========================
 *  Private
 * ========================= */

void SelectDatabaseDialog::setupModel()
{
    m_model->setColumnCount(5);
    m_model->setHorizontalHeaderLabels({
        tr("–ù–∞–∑–≤–∞"),
        tr("Host"),
        tr("Port"),
        tr("–ë–∞–∑–∞"),
        tr("–ö–æ—Ä–∏—Å—Ç—É–≤–∞—á")
    });
}

void SelectDatabaseDialog::loadConnections()
{
    m_model->removeRows(0, m_model->rowCount()); // –û—á–∏—â–µ–Ω–Ω—è —Å—Ç–∞—Ä–∏—Ö –¥–∞–Ω–∏—Ö
    QSettings settings;

    int size = settings.beginReadArray("db_connections");
    for (int i = 0; i < size; ++i) {
        settings.setArrayIndex(i);

        // –°—Ç–≤–æ—Ä—é—î–º–æ —Å–ø–∏—Å–æ–∫ –µ–ª–µ–º–µ–Ω—Ç—ñ–≤ –¥–ª—è –æ–¥–Ω–æ–≥–æ —Ä—è–¥–∫–∞
        QList<QStandardItem*> row {
            new QStandardItem(settings.value("name").toString()),
            new QStandardItem(settings.value("host").toString()),
            new QStandardItem(settings.value("port").toString()),
            new QStandardItem(settings.value("database").toString()),
            new QStandardItem(settings.value("user").toString())
        };

        m_model->appendRow(row); // –¢–µ–ø–µ—Ä –º–æ–¥–µ–ª—å –≤—ñ–¥–ø–æ–≤—ñ–¥–∞—î –∑–∞ –ø–∞–º'—è—Ç—å
    }
    settings.endArray();
}

void SelectDatabaseDialog::saveConnections()
{
    QSettings settings;
    settings.beginWriteArray("db_connections");

    for (int row = 0; row < m_model->rowCount(); ++row) {
        settings.setArrayIndex(row);
        settings.setValue("name",     m_model->item(row, 0)->text());
        settings.setValue("host",     m_model->item(row, 1)->text());
        settings.setValue("port",     m_model->item(row, 2)->text());
        settings.setValue("database", m_model->item(row, 3)->text());
        settings.setValue("user",     m_model->item(row, 4)->text());
    }

    settings.endArray();
}

} // namespace SC::UI::Dialogs
